### PR TITLE
eth_estimateGas update #387

### PR DESCRIPTION
**Pull request by donntrenton**
## PR description
Updated eth_estimateGas description, parameters, and request/return data.

## Fixed Issue(s)
#387 / NC-1899 - eth_estimateGas

### CODE REVIEW

File: docs/Reference/JSON-RPC-API-Methods.md [hunk 1]
```diff
@@ -779,38 +779,75 @@ You can interact with contracts using [eth_sendRawTransaction or eth_call](../Us
...
+Returns an estimate of how much gas is needed for the transaction to complete. The estimate transaction is not added
+to the blockchain. The estimate can be greater than the amount of gas actually used by the transaction, for reasons
+including EVM mechanics and node performance.
 
 **Parameters**
 
 !!!note
-    Parameters are the same as the eth_call parameters, except that all properties are optional. If you do not specify a `gas` limit, Pantheon uses the gas limit from the pending block as an upper bound. As a result, the returned estimate might not be enough to execute the call or transaction when the amount of gas is higher than the pending block's gas limit.
+    The transaction call object parameters are the same as those for eth_call, except that in this method all the
+    fields are optional. If you do not specify a gas limit, Pantheon uses the gas limit from the block at the head
```
[the remark this answers is not in the document]
Comment (MadelineMurray) [reply docs/Reference/JSON-RPC-API-Methods.md]: What happens if you specify a gas limit when calling `eth_estimateGas` that's lower than the estimated gas would be?  Does it return the gas limit you specified when calling `eth_estimateGas` or an estimate that's higher than the specified gas limit?

File: docs/Reference/JSON-RPC-API-Methods.md [hunk 2]
```diff
@@ -779,38 +779,75 @@ You can interact with contracts using [eth_sendRawTransaction or eth_call](../Us
...
+    of the chain as an upper limit. If the amount of gas needed is higher than the estimated upper limit, the call
+    or transaction might not have enough gas to execute.
```
[the remark this answers is not in the document]
Comment (MadelineMurray) [reply docs/Reference/JSON-RPC-API-Methods.md]: I'm not sure it's clear here that `eth_estimateGas` itself doesn't require gas to execute and a seperate call to `eth_sendRawTransaction` is required to actually execute the transaction.

[the remark this answers is not in the document]
Comment (donntrenton) [reply docs/Reference/JSON-RPC-API-Methods.md]: Changed "the call or transaction" to "the transaction"--the original text included "call" and it's just confusing.

Re the second comment thanks for pointing that out--that should be noted. Will add this text to the description: "The eth_estimateGas call requires no gas to execute. You must make a separate call to eth_sendRawTransaction to execute the transaction."

File: docs/Reference/JSON-RPC-API-Methods.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (donntrenton) [reply docs/Reference/JSON-RPC-API-Methods.md]: Researching now where the gas limit referred to is specified.

File: docs/Reference/JSON-RPC-API-Methods.md [hunk 3]
```diff
@@ -779,38 +779,75 @@ You can interact with contracts using [eth_sendRawTransaction or eth_call](../Us
...
-Generates and returns an estimate of how much gas is necessary to allow the transaction to complete. (Per Etherscan: gas price * gas used.) The transaction is added to the blockchain. The estimate may be significantly more than the amount of gas actually used by the transaction for reasons including EVM mechanics and node performance.
+Returns an estimate of how much gas is needed for the transaction to complete. The estimate transaction is not added
+to the blockchain. The estimate can be greater than the amount of gas actually used by the transaction, for reasons
+including EVM mechanics and node performance.
 
 **Parameters**
 
 !!!note
-    Parameters are the same as the eth_call parameters, except that all properties are optional. If you do not specify a `gas` limit, Pantheon uses the gas limit from the pending block as an upper bound. As a result, the returned estimate might not be enough to execute the call or transaction when the amount of gas is higher than the pending block's gas limit.
+    The transaction call object parameters are the same as those for eth_call, except that in this method all the
```
[the remark this answers is not in the document]
Comment (donntrenton) [reply docs/Reference/JSON-RPC-API-Methods.md]: Fixed. I applied backquotes and links where needed.

File: docs/Reference/JSON-RPC-API-Methods.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (donntrenton) [reply docs/Reference/JSON-RPC-API-Methods.md]: The Transaction Call Object has a gas field; eth_estimateGas also uses this object, so I believe this is how you specify the gas amount referred to. The Block Object has a gasLimit field--that's the value Pantheon uses to estimate an upper limit. So I updated the text as follows:

Change:
"If you do not specify a gas limit, Pantheon uses the gas limit from the block at the head of the chain as an upper limit."
To:
"If you do not specify a `gas` amount in the transaction call object, Pantheon uses the
    `gasLimit` amount from the block at the head of the chain as the upper limit."

File: docs/Reference/JSON-RPC-API-Methods.md [hunk 4]
```diff
@@ -779,38 +779,75 @@ You can interact with contracts using [eth_sendRawTransaction or eth_call](../Us
...
 *OBJECT* - [Transaction call object](JSON-RPC-API-Objects.md#transaction-call-object).
 
-*QUANTITY|TAG* - Integer representing a block number or one of the string tags `latest`, `earliest`, or `pending`, as described in [Block Parameter](Using-JSON-RPC-API.md#block-parameter).
-
 **Returns**
 
-`result` (*QUANTITY*) -  Amount of gas used.
+`result` : `quantity` -  Amount of gas used.
+
+The following returns an estimate of 21000 wei (0x5208) for the transaction.
```
[the remark this answers is not in the document]
Comment (donntrenton) [reply docs/Reference/JSON-RPC-API-Methods.md]: To clarify, there are two separate examples in the eth_estimateGas section. The first one was data from my call using the quickstart private network, and the second one is an example that Lucas provided. We don't currently use subtitles to separate examples e.g. Example 1 and Example 2. Hope that was clear.

[the remark this answers is not in the document]
Comment (MadelineMurray) [reply docs/Reference/JSON-RPC-API-Methods.md]: Got it now.

File: docs/Reference/JSON-RPC-API-Methods.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (MadelineMurray) [reply docs/Reference/JSON-RPC-API-Methods.md]: I'm still not clear on how the gas limit relates to the calculated estimate:

What happens if you specify a gas limit when calling eth_estimateGas that's lower than the estimated gas would be? Does it return the gas limit you specified when calling eth_estimateGas or an estimate that's higher than the specified gas limit?

[the remark this answers is not in the document]
Comment (donntrenton) [reply docs/Reference/JSON-RPC-API-Methods.md]: Note we had a long thread in Slack on this issue concluding with Madeline saying: "as your testing confirms, the gas limit (whether set on the transaction call object or not and regardless of what it is at the head of the chain), for eth_estimateGas is irrelevant to what it returns. Which I think is what we need to communicate."

So I'd remove "If you do not specify a `gas` amount in the transaction call object" and state only "Pantheon uses the gas limit from the block at the head etc."

[the remark this answers is not in the document]
Comment (donntrenton) [reply docs/Reference/JSON-RPC-API-Methods.md]: I'm going to split off this issue in a separate issue so we can close the PR.

**Pull request merged**